File: src/main/java/org/lanternpowered/server/block/BlockTypeBuilder.java
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.block;

import org.lanternpowered.server.behavior.Behavior;
import org.lanternpowered.server.behavior.pipeline.BehaviorPipeline;
import org.lanternpowered.server.behavior.pipeline.MutableBehaviorPipeline;
import org.lanternpowered.server.block.provider.BlockObjectProvider;
import org.lanternpowered.server.block.provider.property.PropertyProviderCollection;
import org.lanternpowered.server.item.ItemTypeBuilder;
import org.spongepowered.api.block.BlockState;
import org.spongepowered.api.block.tileentity.TileEntity;
import org.spongepowered.api.block.tileentity.TileEntityType;
import org.spongepowered.api.block.trait.BlockTrait;
import org.spongepowered.api.text.translation.Translation;
import org.spongepowered.api.util.AABB;

import java.util.Collection;
import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Supplier;

import org.checkerframework.checker.nullness.qual.Nullable;

public interface BlockTypeBuilder {

    BlockTypeBuilder selectionBox(@Nullable AABB selectionBox);

    BlockTypeBuilder selectionBox(@Nullable Function<BlockState, AABB> selectionBoxProvider);

    BlockTypeBuilder selectionBox(@Nullable BlockObjectProvider<AABB> selectionBoxProvider);

    BlockTypeBuilder collisionBox(@Nullable AABB collisionBox);

    BlockTypeBuilder collisionBox(@Nullable Function<BlockState, AABB> collisionBoxProvider);

    BlockTypeBuilder collisionBox(@Nullable BlockObjectProvider<AABB> collisionBoxProvider);

    BlockTypeBuilder collisionBoxes(@Nullable Collection<AABB> collisionBoxes);

    BlockTypeBuilder collisionBoxes(@Nullable Function<BlockState, Collection<AABB>> collisionBoxesProvider);

    BlockTypeBuilder collisionBoxes(@Nullable BlockObjectProvider<Collection<AABB>> collisionBoxesProvider);

    BlockTypeBuilder defaultState(Function<BlockState, BlockState> function);

    BlockTypeBuilder properties(PropertyProviderCollection collection);

    BlockTypeBuilder properties(Consumer<PropertyProviderCollection.Builder> consumer);

    BlockTypeBuilder tileEntityType(Supplier<TileEntityType> tileEntityType);

    BlockTypeBuilder tileEntity(Supplier<TileEntity> supplier);

    BlockTypeBuilder tileEntity(BlockEntityProvider tileEntityProvider);

    BlockTypeBuilder traits(BlockTrait<?>... blockTraits);

    BlockTypeBuilder trait(BlockTrait<?> blockTrait);

    BlockTypeBuilder translation(String translation);

    BlockTypeBuilder translation(Translation translation);

    BlockTypeBuilder translation(TranslationProvider translationProvider);

    BlockTypeBuilder behaviors(BehaviorPipeline<Behavior> behaviorPipeline);

    BlockTypeBuilder behaviors(Consumer<MutableBehaviorPipeline<Behavior>> consumer);

    BlockTypeBuilder itemType();

    BlockTypeBuilder itemType(Consumer<ItemTypeBuilder> consumer);

    LanternBlockType build(String pluginId, String id);
}

File: src/main/kotlin/org/lanternpowered/server/block/BlockTypeBuilder.kt
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.block

import org.lanternpowered.api.block.BlockState
import org.lanternpowered.api.block.BlockType
import org.lanternpowered.api.key.NamespacedKey
import org.lanternpowered.api.key.namespacedKey
import org.lanternpowered.api.text.Text
import org.lanternpowered.api.util.AABB
import org.lanternpowered.api.util.math.times
import org.lanternpowered.server.behavior.Behavior
import org.lanternpowered.server.behavior.pipeline.MutableBehaviorPipeline
import org.lanternpowered.server.block.property.FlammableInfo
import org.lanternpowered.server.block.state.BlockStateProperties
import org.lanternpowered.server.data.LocalKeyRegistry
import org.lanternpowered.server.data.key.LanternKeys
import org.lanternpowered.server.item.ItemTypeBuilder
import org.spongepowered.api.block.entity.BlockEntity
import org.spongepowered.api.block.entity.BlockEntityType
import org.spongepowered.api.block.entity.BlockEntityTypes
import org.lanternpowered.api.data.Keys
import org.spongepowered.api.state.StateProperty
import org.spongepowered.math.vector.Vector3d
import java.util.function.Supplier

val testBlockType = blockTypeOf(namespacedKey("namespace", "value")) {
    name("Test Block")
    stateProperty(BlockStateProperties.IS_WET, false)
    keys {
        register(Keys.BLOCK_SOUND_GROUP, BlockSoundGroups.GLASS)
        register(Keys.BLAST_RESISTANCE, 10.2)
    }
    stateKeys {
        registerProvider(LanternKeys.FLAMMABLE_INFO) {
            get {
                if (this.getStateProperty(BlockStateProperties.IS_WET).orElse(false)) {
                    null
                } else {
                    FlammableInfo(1, 1)
                }
            }
            getDirectional {
                FlammableInfo(1, 1)
            }
        }
    }
    blockEntity(BlockEntityTypes.BANNER)
    collisionBox {
        if (getStateProperty(BlockStateProperties.IS_WET).orElse(false)) {
            AABB(Vector3d.ZERO, Vector3d.ONE * 2)
        } else {
            AABB(Vector3d.ZERO, Vector3d.ONE)
        }
    }
    itemType()
    itemType {
        name("Something else")
    }
    behaviors {

    }
}

fun blockTypeOf(key: NamespacedKey, fn: BlockTypeBuilder.() -> Unit): BlockType {
    TODO()
}

@Target(AnnotationTarget.TYPE, AnnotationTarget.CLASS)
@DslMarker
annotation class BlockTypeBuilderDsl

@BlockTypeBuilderDsl
interface BlockTypeBuilder {

    fun name(name: String)
    fun name(name: Text)

    /**
     * Adds a [StateProperty] with the given default value.
     */
    fun <V : Comparable<V>> stateProperty(stateProperty: StateProperty<V>, defaultValue: V)

    /**
     * Applies a [BlockEntity] to the block type.
     */
    fun blockEntity(blockEntityType: BlockEntityType)

    /**
     * Applies a [BlockEntity] to the block type.
     */
    fun blockEntity(blockEntityType: Supplier<out BlockEntityType>) = this.blockEntity(blockEntityType.get())

    /**
     * Applies the selection bounding box.
     */
    fun selectionBox(selectionBox: AABB?)

    /**
     * Applies the selection bounding box based on the [BlockState].
     */
    fun selectionBox(fn: @BlockTypeBuilderDsl BlockState.() -> AABB)

    /**
     * Applies the collision bounding box.
     */
    fun collisionBox(collisionBox: AABB?)

    /**
     * Applies the collision bounding box based on the [BlockState].
     */
    fun collisionBox(fn: @BlockTypeBuilderDsl BlockState.() -> AABB)

    /**
     * Applies the collision bounding boxes.
     */
    fun collisionBoxes(collisionBox: Collection<AABB>)

    /**
     * Applies the collision bounding boxes based on the [BlockState].
     */
    fun collisionBoxes(fn: @BlockTypeBuilderDsl BlockState.() -> Collection<AABB>)

    /**
     * Applies keys to the [BlockType].
     */
    fun keys(fn: @BlockTypeBuilderDsl LocalKeyRegistry<BlockType>.() -> Unit)

    /**
     * Applies keys to the [BlockState]. Overrides keys provided through [keys].
     */
    fun stateKeys(fn: @BlockTypeBuilderDsl LocalKeyRegistry<BlockState>.() -> Unit)

    /**
     * Applies behaviors to the [BlockType].
     */
    fun behaviors(fn: @BlockTypeBuilderDsl MutableBehaviorPipeline<Behavior>.() -> Unit)

    /**
     * Enables a item type for the block type and allows the item type to be modified.
     */
    fun itemType(fn: @BlockTypeBuilderDsl ItemTypeBuilder.() -> Unit = {})

}
